File: smsly-core/smsly_core/adapters.py
"""
Unified Provider Adapter Module
================================
Base classes for SMS/MMS/WhatsApp/RCS provider integrations.
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Optional, Dict, Any, List
from enum import Enum
import structlog

logger = structlog.get_logger(__name__)


class MessageStatus(str, Enum):
    PENDING = "pending"
    SENT = "sent"
    DELIVERED = "delivered"
    FAILED = "failed"
    REJECTED = "rejected"


@dataclass
class SendResult:
    """Result of a message send operation."""
    success: bool
    provider_message_id: Optional[str] = None
    status: MessageStatus = MessageStatus.PENDING
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    raw_response: Optional[Dict[str, Any]] = None
    cost: Optional[float] = None
    segments: int = 1


@dataclass
class WebhookEvent:
    """Parsed webhook event from a provider."""
    provider_message_id: str
    status: MessageStatus
    timestamp: Optional[float] = None
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    raw_payload: Optional[Dict[str, Any]] = None


class BaseProviderAdapter(ABC):
    """
    Abstract base class for messaging provider adapters.
    
    All provider implementations (Twilio, Vonage, etc.) should inherit from this.
    """
    
    name: str = "base"
    supports_mms: bool = False
    supports_whatsapp: bool = False
    supports_rcs: bool = False
    
    def __init__(self, config: Dict[str, Any]):
        """
        Initialize the adapter with provider-specific configuration.
        
        Args:
            config: Provider-specific config (API keys, account IDs, etc.)
        """
        self.config = config
        self._is_initialized = False
    
    async def initialize(self) -> None:
        """Initialize the adapter (e.g., create HTTP clients)."""
        self._is_initialized = True
        logger.info(f"Provider adapter initialized", provider=self.name)
    
    async def close(self) -> None:
        """Clean up resources (e.g., close HTTP clients)."""
        self._is_initialized = False
        logger.info(f"Provider adapter closed", provider=self.name)
    
    @abstractmethod
    async def send_sms(
        self,
        to: str,
        from_: str,
        body: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> SendResult:
        """
        Send an SMS message.
        
        Args:
            to: Recipient phone number (E.164 format)
            from_: Sender phone number or alphanumeric ID
            body: Message content
            metadata: Optional metadata for tracking
        
        Returns:
            SendResult with provider response
        """
        pass
    
    async def send_mms(
        self,
        to: str,
        from_: str,
        text: Optional[str],
        media_urls: List[str],
        metadata: Optional[Dict[str, Any]] = None,
    ) -> SendResult:
        """
        Send an MMS message.
        
        Default implementation raises NotImplementedError.
        Override in providers that support MMS.
        """
        raise NotImplementedError(f"{self.name} does not support MMS")
    
    async def validate_webhook(
        self,
        headers: Dict[str, str],
        body: bytes,
    ) -> bool:
        """
        Validate webhook signature from provider.
        
        Args:
            headers: HTTP headers from webhook request
            body: Raw request body
        
        Returns:
            True if signature is valid
        """
        return True  # Override in providers with signature verification
    
    async def parse_webhook(self, body: bytes) -> WebhookEvent:
        """
        Parse webhook payload into a standardized WebhookEvent.
        
        Args:
            body: Raw request body
        
        Returns:
            Parsed WebhookEvent
        """
        raise NotImplementedError(f"{self.name} must implement webhook parsing")
    
    async def health_check(self) -> bool:
        """
        Check if the provider is reachable.
        
        Returns:
            True if provider API is healthy
        """
        return self._is_initialized


class ProviderRegistry:
    """Registry for managing multiple provider adapters."""
    
    def __init__(self):
        self._adapters: Dict[str, BaseProviderAdapter] = {}
    
    def register(self, adapter: BaseProviderAdapter) -> None:
        """Register a provider adapter."""
        self._adapters[adapter.name.lower()] = adapter
        logger.info("Provider registered", provider=adapter.name)
    
    def get(self, name: str) -> BaseProviderAdapter:
        """Get a provider adapter by name."""
        adapter = self._adapters.get(name.lower())
        if not adapter:
            raise ValueError(f"Unknown provider: {name}")
        return adapter
    
    def list(self) -> List[str]:
        """List all registered provider names."""
        return list(self._adapters.keys())
    
    async def close_all(self) -> None:
        """Close all registered adapters."""
        for adapter in self._adapters.values():
            await adapter.close()


# Global registry instance
provider_registry = ProviderRegistry()

File: smsly-rust/smsly-core/src/adapters.rs
use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use thiserror::Error;
use tokio::sync::RwLock;
use tracing::info;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageStatus {
    Pending,
    Sent,
    Delivered,
    Failed,
    Rejected,
}

impl Default for MessageStatus {
    fn default() -> Self {
        Self::Pending
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SendResult {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_message_id: Option<String>,
    pub status: MessageStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub raw_response: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost: Option<f64>,
    pub segments: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebhookEvent {
    pub provider_message_id: String,
    pub status: MessageStatus,
    pub timestamp: Option<f64>,
    pub error_code: Option<String>,
    pub error_message: Option<String>,
    pub raw_payload: Option<Value>,
}

#[derive(Error, Debug)]
pub enum AdapterError {
    #[error("Unknown provider: {0}")]
    UnknownProvider(String),
}

#[async_trait]
pub trait BaseProviderAdapter: Send + Sync {
    fn name(&self) -> String;
    fn supports_mms(&self) -> bool {
        false
    }
    fn supports_whatsapp(&self) -> bool {
        false
    }
    fn supports_rcs(&self) -> bool {
        false
    }

    async fn initialize(&self) {
        info!("Provider adapter initialized: {}", self.name());
    }

    async fn close(&self) {
        info!("Provider adapter closed: {}", self.name());
    }

    async fn send_sms(
        &self,
        to: &str,
        from: &str,
        body: &str,
        metadata: Option<HashMap<String, Value>>,
    ) -> SendResult;

    async fn send_mms(
        &self,
        _to: &str,
        _from: &str,
        _text: Option<&str>,
        _media_urls: Vec<String>,
        _metadata: Option<HashMap<String, Value>>,
    ) -> SendResult {
        // Default implementation for providers that don't support MMS
        SendResult {
            success: false,
            status: MessageStatus::Failed,
            error_message: Some(format!("{} does not support MMS", self.name())),
            ..Default::default()
        }
    }

    async fn validate_webhook(&self, _headers: &HashMap<String, String>, _body: &[u8]) -> bool {
        true
    }

    async fn parse_webhook(&self, _body: &[u8]) -> Result<WebhookEvent, String> {
        Err(format!("{} must implement webhook parsing", self.name()))
    }

    async fn health_check(&self) -> bool {
        true
    }
}

pub struct ProviderRegistry {
    adapters: RwLock<HashMap<String, Arc<Box<dyn BaseProviderAdapter>>>>,
}

impl ProviderRegistry {
    pub fn new() -> Self {
        Self {
            adapters: RwLock::new(HashMap::new()),
        }
    }

    pub async fn register(&self, adapter: Box<dyn BaseProviderAdapter>) {
        let name = adapter.name().to_lowercase();
        info!("Provider registered: {}", name);
        self.adapters.write().await.insert(name, Arc::new(adapter));
    }

    pub async fn get(&self, name: &str) -> Result<Arc<Box<dyn BaseProviderAdapter>>, AdapterError> {
        let adapters = self.adapters.read().await;
        adapters
            .get(&name.to_lowercase())
            .cloned()
            .ok_or_else(|| AdapterError::UnknownProvider(name.to_string()))
    }

    pub async fn list(&self) -> Vec<String> {
        self.adapters.read().await.keys().cloned().collect()
    }
}
